### DISCUSSION TITLE
Make parse tree serializable and interchangeable between runtimes

### DISCUSSION THREAD

**Discussion by dimmik** [antlr/antlr4#4680] [Ideas]
**Scenario:** 
Python runtime.
Relatively complex language grammar, and really complex translation code (from language to ASM).

The grammar is parsed in several seconds in python. 
Even though translation code is complex, the translation runs in ~0.3s in python.
So, tree parsing is consuming ~90% of processing time.
(tried) The same grammar parsing is 3 times faster in c#, and (guess, didn't really try) 10+x faster in C.

Devs, implementing the translation code, are familiar with python, not C/C#.

**Suggestion:**
It would have been nice if the parse tree, that is result of parsing, was produced by C code, then serialized, then loaded into python code, and then used for translation.
That requires the parse tree to be serializable, and understandable by both (ideally all) runtimes.

That would unblock both performance (that is way better in C) and dev speed (that is better in python).

**Comment by jimidle** [comment]
Are you translating directly from the parse tree? It would be much easier
for you to generate an ast and serialize that. Then you have full control
of it all.

Also, how efficient is your grammar? Tree walking should be fast, but if
your grammar is ambiguous then you well lose performance there.

Python isn’t the correct choice for a fast parse though, as you have
noticed.

**Reply by dimmik** [reply]
Definitely anything can be implemented, and any problem can be addressed in this or another way.
Request is not about my problem, request is about making parse tree serializable.
The scenario is just an illustration of when it might be useful.

**Comment by jimidle** [comment]
It is unlikely to be implemented. While the data may be the same, the
generators do things differently, including static initialization. Take a
look at all the different generated code. This seems like an XY question
